### DISCUSSION TITLE
Usage of insert

### DISCUSSION THREAD

**Discussion by AntonioSun** [smicyk/groovy-jmeter#101] [Q&A] [answered]
Hi, 

I vaguely remember that the fragment for `insert` can only contains a single element, but cannot find where I read it now. 

So, let me confirm it here, can I have a insert file of structure:

```
fragment {
    http(url1) 
    http(url2) 
    ...
    http(urln) 
}
```
?

thx

**Comment by AntonioSun** [comment]
> I vaguely remember that the fragment for `insert` can only contains a single element

Not true. The example in OP works perfectly.

Maybe what I remember was include...

**Reply by AntonioSun** [reply]
Nope. I concluded too soon.

So I changed the file

https://github.com/smicyk/groovy-jmeter/blob/b5887dd3b602da188bfaa6c7b1cb50acb45b3df8/examples/structure/fragments/login.groovy#L1-L10

To have one more http req (2 http reqs in total), and it worked fine. (Hence I concluded prematurely). However, 

If I add two more http reqs (3 http reqs in total), then it fails, and only single http req showed up in the `--jmx-out`.

Here is my test file:

```
fragment {
    http('POST /login') {
        params {
            param(name: 'username', value: 'john')
            param(name: 'password', value: 'john')
        }

        extract_regex expression: '"trackId", content="([0-9]+)"', variable: 'var_trackId'
    }

    http('GET /api/books') {
	params values: [ limit: '10' ]

	extract_json expressions: '$..id', variables: 'var_bookId'
    }
    http('POST /login') {
        params {
            param(name: 'username', value: 'john')
            param(name: 'password', value: 'john')
        }

        extract_regex expression: '"trackId", content="([0-9]+)"', variable: 'var_trackId'
    }
}
```

and here is its --jmx-out:

<img width="202" alt="image" src="https://user-images.githubusercontent.com/2840074/180617686-46cb30f3-d6ce-4261-ad2d-dea53a985ee3.png">

Moreover, if my fragment file is as complicated as the following, then the `--jmx-out` would be totally wrong:

```
fragment {

      defaults(protocol: '${c_app_protocol}', domain: '${c_app_host_name}', port:  '${c_app_host_port}')
      headers {
        header(name: 'Host', value: '${c_app_host_name}')
        header(name: 'Origin', value: '${c_app_host_name}')
        header(name: 'Referer', value: '${c_app_host_name}')
        header(name: 'Connection', value: 'keep-alive')
        header(name: 'Cache-Control', value: 'max-age=0')
        header(name: 'Upgrade-Insecure-Requests', value: '1')
        header(name: 'User-Agent', value: 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/102.0.0.0 Safari/537.36')
      }
      cookies()
      cache()
      // check_response {
      //   status() eq 200
      // }
      check_response applyTo: 'parent', {
        text() excludes '${c_app_error_kw}'
      }

      transaction('T0x_/') {
        http (method: 'GET', path: '/', name: 'T0r_/') {
          headers {
            header(name: 'Accept', value: 'text/html, application/xhtml+xml')
          }
        }
      }

      transaction('T0x_A') {
        http (method: 'GET', path: '/A', name: 'T0r_/') {
          headers {
            header(name: 'Accept', value: 'text/html, application/xhtml+xml')
          }
        }
      }

      transaction('T0x_B') {
        http (method: 'GET', path: '/B', name: 'T0r_/') {
          headers {
            header(name: 'Accept', value: 'text/html, application/xhtml+xml')
          }
        }
      }
   . . .
```

**Reply by smicyk** [reply]
Hi, with this fix #102 I add some improvments regarding insert element. Not it can be used with multiple elements inside fragment. It also can be put anywhere in the hierarchy.

**Comment by AntonioSun** [comment, marked as answer]
> Hi, with this fix #102 I add some improvments regarding insert element. Now it can be used with multiple elements inside fragment. It also can be put anywhere in the hierarchy.

Bravo!!! 
This is super cool!
